### PR TITLE
Limit multi-stage agg alternatives generation for DQAs

### PR DESCRIPTION
**Pull request by melanieplageman**
For a query containing a distinct qualified aggregate (DQA) on the distribution column, such as the below query,
```
create table foo (a int, b int) distributed by (a);
explain select count(distinct a) from foo;
```
the plan chosen was a global one-stage aggregate:
```
                                     QUERY PLAN
-----------------------------------------------------------------------------------------
 Aggregate  (cost=0.00..431.97 rows=1 width=8)
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..431.92 rows=100478 width=4)
         ->  Table Scan on foo  (cost=0.00..431.02 rows=33493 width=4)
 Settings:  optimizer=on; optimizer_segments=128
 Optimizer status: PQO version 2.51.3
(5 rows)
```
This plan is inefficient, especially in cases with few distinct values in each segment. The better plan is to do a local aggregate before gathering to the master. This plan alternative was being eliminated during logical transform.

Since a multi-stage agg is preferrable and we have a strong preference for three-stage aggregates, previously,only a three-stage alternative was generated by the transform but prohibited during property enforcement when the DQA was on the distribution column.
With this PR, we are generating both alternatives (two-stage as well as three-stage) and leaving the elimination to the optimization phase.

Also, as part of this PR, we are eliminating generating multi-stage agg alternatives for queries containing a non-scalar DQA(Distinct Qualified Aggregate) on the distribution column as well as non-splittable aggregate functions.

### CODE REVIEW

File: libgpopt/src/xforms/CXformSplitDQA.cpp [hunk 1]
```diff
@@ -157,61 +157,86 @@ CXformSplitDQA::Transform
...
 	}
 
-	if (fScalarDQA && !fForce3StageScalarDQA) {
-		// if only want 3-stage DQA then skip this 2-stage option for scalar DQA.
+	CExpression *pexprPrEl = (*pexprProjectList)[0];
+	CExpression *pexprAggFunc = (*pexprPrEl)[0];
+	CScalarAggFunc *popAggfunc = CScalarAggFunc::PopConvert(pexprAggFunc->Pop());
+	/*
+	* If the AggFunc is not splittable and the group by col
+	* is not contained in the distribution columns,
```
Comment (oarap) [inline libgpopt/src/xforms/CXformSplitDQA.cpp]: not contained or not equal to? I think you need to change the comment.

Comment (khannaekta) [reply libgpopt/src/xforms/CXformSplitDQA.cpp]: Updated the comment

File: libgpopt/src/xforms/CXformSplitDQA.cpp [hunk 2]
```diff
@@ -157,61 +157,86 @@ CXformSplitDQA::Transform
...
+	CExpression *pexprPrEl = (*pexprProjectList)[0];
+	CExpression *pexprAggFunc = (*pexprPrEl)[0];
+	CScalarAggFunc *popAggfunc = CScalarAggFunc::PopConvert(pexprAggFunc->Pop());
+	/*
+	* If the AggFunc is not splittable and the group by col
+	* is not same as the distribution columns,
+	* no 3 stage/2 stage agg alternatives are not required
+	*/
+	if (pmda->Pmdagg(popAggfunc->Pmdid())->FSplittable() && fGBColNotDistribCol)
+	{
```
Comment (d) [inline libgpopt/src/xforms/CXformSplitDQA.cpp]: What test would break if you were to remove the check `pmda->Pmdagg(popAggfunc->Pmdid())->FSplittable()`?

It appears to me that this "splittable" check is redundant: if the input aggregate function is not splittable, execution of `CXformSplitDQA::Transform` won't even reach to this point, because `ExtractDistinctCols` would have left an empty `pdrgpcrArgDQA` a few lines above here.

File: libgpopt/src/xforms/CXformSplitDQA.cpp [hunk 3]
```diff
@@ -157,61 +157,86 @@ CXformSplitDQA::Transform
...
-				pmda,
-				pexpr,
-				pexprRelational,
-				phmexprcr,
-				pdrgpcrArgDQA
-				);
-
-		pxfres->Add(pexprAlt2);
+	if(pexprRelational->Pop()->Eopid() == COperator::EopLogicalGet)
+	{
```
Comment (d) [inline libgpopt/src/xforms/CXformSplitDQA.cpp]: The code suggests the intent here is to check for how the relational child is distributed. If so, it's *wrong*. Because:

1. An exploration xform is now overreaching into the enforcement / optimization phase: just because we are on aggregating on top of a "get" doesn't mean there will never be motion in between us (ok sort of yes, but that assumption is shaky).
1. It's fragile: what if there is a simple filter (CLogicalSelect) over the logical get? Do we also peek under a "Logical Select"? How about "CLogicalProject"?

This is precisely the kind of things that ORCA's architecture is good for: generate pure logical alternatives, implement them, "flesh them out" during enforcement, then (finally) pick the cheapest alternative.

**Comment by d** [comment]
TL;DR: It's my end of day, so I'm asking to confirm my learning of this PR ...

if we do not force-pick 3-phase scalar DQA aggs (`set optimizer_force_three_stage_scalar_dqa=on`, or `gporca_test -T 10405`), the split-agg alternative would be:

```
Agg
  Output: aggfn(final a) [global]
  -> Agg
       Output: aggfn(partial DISTINCT a) [local]
       -> Get (foo)
            Output: a
```

I didn't spend much time looking over the change, but my takeaway is this PR adds a new logical alternative for `SELECT aggfn(DISTINCT a) FROM foo)`:

```
Agg
  Output: aggfn(a)
  ->  Group By (a) [local]
        Output: a
	-> Get (foo)
	   Output: a
```

Am I reading your code right?

Jesse

File: libgpopt/src/xforms/CXformSplitDQA.cpp (hunk 2 above)
Comment (khannaekta) [reply libgpopt/src/xforms/CXformSplitDQA.cpp]: Thanks for pointing it out. Will remove that check.

File: libgpopt/src/xforms/CXformSplitDQA.cpp (hunk 3 above)
Comment (khannaekta) [reply libgpopt/src/xforms/CXformSplitDQA.cpp]: For the first point, in `CPhysicalAgg::EpetDistribution` we see that the Property enforcement takes care to prohibit plans if local aggregate already delivers the enforced distribution. Thus, we can go ahead with the assumption.

For the Second point, the code added as part of the PR is only for cases specific to queries that have a DQA on the distribution column. In all other cases, the xform will generate results as of today.

Agreed to your point, that ORCA's  architecture is good for generating logical alternatives, implementing them, and then finally pick the cheapest alternative(after enforcing properties).
But in case the query is a DQA on the distribution column, the xform did not even generate alternative for a 2-stage agg. This PR, takes care of going ahead with generating such logical alternatives which would be picked up later by ORCA during optimization.

**Comment by khannaekta** [comment]
@d yes your understanding of the code is correct. This PR goes ahead with generating the logical alternative(which previously wasn't generated due to the GUC `optimizer_force_three_stage_scalar_dqa`) for `SELECT aggfn(DISTINCT a) FROM foo`:
```
Agg
  Output: aggfn(normal a)
  ->  Group By (a) [local dedupe]
        Output: a
	-> Get (foo)
	   Output: a
```

File: libgpopt/src/xforms/CXformSplitDQA.cpp (hunk 3 above)
Comment (hsyuan) [reply libgpopt/src/xforms/CXformSplitDQA.cpp]: it would be too specific for logical get. it can be a join too.

File: data/dxl/minidump/UDA-Distinct.mdp [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+language sql immutable;
+
+create aggregate aggfns(integer,integer,text) (
+   sfunc = aggfns_trans, stype = aggtype[],
+   initcond = '{}'
+);
+
+create table jazz(a int, b int, c text);
+explain select aggfns(distinct a,b,c) from jazz;
+-->
```
Comment (hsyuan) [inline data/dxl/minidump/UDA-Distinct.mdp]: The query in this minidump is not enough to help test your code changes. It only generate single stage agg, what about 2 stage agg for splittable DQA?

File: libgpopt/src/xforms/CXformSplitDQA.cpp [hunk 5]
```diff
@@ -157,61 +157,86 @@ CXformSplitDQA::Transform
...
-        
-	pxfres->Add(pexprAlt1);
-
 	DrgPcr *pDrgPcr = CLogicalGbAgg::PopConvert(pexpr->Pop())->Pdrgpcr();
 	BOOL fScalarDQA = (pDrgPcr == NULL || pDrgPcr->UlLength() == 0);
 	BOOL fForce3StageScalarDQA = GPOS_FTRACE(EopttraceForceThreeStageScalarDQA);
-	if (!(fForce3StageScalarDQA && fScalarDQA)) {
-		// we skip this option if it is a Scalar DQA and we only want plans with 3-stages of aggregation
+	BOOL fDQAColDistribCol = false;
+	BOOL fGBColNotDistribCol = true;
```
Comment (hsyuan) [inline libgpopt/src/xforms/CXformSplitDQA.cpp]: I don't understand what these vars mean here.

Comment (khannaekta) [reply libgpopt/src/xforms/CXformSplitDQA.cpp]: `fDQAColDistribCol` - Flag to specify if DQA column is same as the distribution column
`fGBColNotDistribCol` - Flag to specify if group by column is not same as the distribution column

`fDQAColDistribCol` should always be set to false unless we have the DQA column same as that of the distribution column. This flag is checked for generating the 2-stage alternative.

`fGBColNotDistribCol` should always be set to true as we want to explore the alternatives of this Xform. Only if the group by column is same as the distribution column, we do not want to explore the multi stage aggs as it can be achieved by a local aggregation (generated as part of CXformSimplifyGbAgg)

File: libgpopt/src/xforms/CXformSplitDQA.cpp (hunk 3 above)
Comment (khannaekta) [reply libgpopt/src/xforms/CXformSplitDQA.cpp]: Agreed, it is for a specific case of logical get. This PR is targeted only to a specific case of DQA on the distribution column.
For all other cases it goes ahead with what it was doing today(not causing any regression). So in case of a join/partition table, it will go ahead for generating a 3-stage agg which is finally picked up later by ORCA during optimization.
But for a query of the form `select agg_fn(distinct a) from foo;` where a is the distribution column, a 3-stage agg is discarded by the property enforcement framework in the optimization phase.

File: data/dxl/minidump/UDA-Distinct.mdp (hunk 4 above)
Comment (khannaekta) [reply data/dxl/minidump/UDA-Distinct.mdp]: Those tests are already part of `CAggTest`.  We did not have any test for a UDF which is not splittable, thus added it. Will still double check.

File: libgpopt/src/xforms/CXformSplitDQA.cpp (hunk 3 above)
Comment (hsyuan) [reply libgpopt/src/xforms/CXformSplitDQA.cpp]: For this query, `select count(distinct a) from (select r1.a from r1 left join s1 on r1.a = s1.a ) r1` (both r1 and s1 are distributed by a), it should generate 2 stage agg, but orca still generate single stage agg. So does for partition table: `select count(distinct a) from pt` (pt is distributed by a, partition by b);

The PR should not be targeted only to the specific case, it should apply to other cases too.

File: libgpopt/src/xforms/CXformSplitDQA.cpp (hunk 5 above)
Comment (hsyuan) [reply libgpopt/src/xforms/CXformSplitDQA.cpp]: add this to the code's comment.

File: data/dxl/minidump/UDA-Distinct.mdp (hunk 4 above)
Comment (hsyuan) [reply data/dxl/minidump/UDA-Distinct.mdp]: Do you have query in minidump that mimics customer's query?

Comment (khannaekta) [reply data/dxl/minidump/UDA-Distinct.mdp]: Yes, the minidump `data/dxl/minidump/ThreeStageAgg-ScalarAgg-DistinctDistrCol.mdp` mimics customer query. Previously the minidump did not have the ORCA traceflag `EopttraceForceThreeStageScalarDQA` (104005), thus it was generating a plan with two-stage (Therefore, no plan change diff).
Updated the minidump to have this traceflag as it is default in GPDB.

Comment (hsyuan) [reply data/dxl/minidump/UDA-Distinct.mdp]: ok, cool

File: libgpopt/src/xforms/CXformSplitDQA.cpp (hunk 3 above)
Comment (dhanashreek89) [reply libgpopt/src/xforms/CXformSplitDQA.cpp]: can you please also check if the problem exists for partitioned table as well? In which case we may want to extend this check for CLogicalDynamicGet.

**Comment by khannaekta** [comment]
This PR aims at a very specific case of the problem. We need to rethink the solution. 
Closing this PR for now.
